src/circuit_builder.rs

```diff
@@ -10,111 +10,64 @@ use crate::commitment_scheme::kzg10::PublicParameters;
-    pub fn to_bytes(&self) -> [u8; Self::serialized_size()] {
-        let mut bytes = [0u8; Self::serialized_size()];
-        match self {
-            Self::BlsScalar(scalar, _) => {
-                bytes[0] = BLS_SCALAR;
-                bytes[1..33].copy_from_slice(&scalar.to_bytes());
-                bytes
-            }
-            Self::JubJubScalar(scalar, _) => {
-                bytes[0] = JUBJUB_SCALAR;
-                bytes[1..33].copy_from_slice(&scalar.to_bytes());
-                bytes
-            }
-            Self::AffinePoint(point, _, _) => {
-                bytes[0] = JUBJUB_AFFINE;
-                bytes[1..Self::serialized_size()].copy_from_slice(&point.to_bytes());
-                bytes
-            }
-        }
+impl From<BlsScalar> for PublicInputValue {
+    fn from(scalar: BlsScalar) -> Self {
+        Self(vec![scalar])
     }
+}
 
-    /// Generate a [`PublicInput`] structure from it's byte representation.
-    pub fn from_bytes(bytes: &[u8]) -> Result<Self, Error> {
-        if bytes.len() < Self::serialized_size() {
-            return Err(Error::InvalidPublicInputBytes);
-        } else {
-            let mut array_bytes = [0u8; 32];
-            array_bytes.copy_from_slice(&bytes[1..Self::serialized_size()]);
-            match bytes[0] {
-                BLS_SCALAR => BlsScalar::from_bytes(&array_bytes)
-                    .map(|s| Self::BlsScalar(s, 0))
-                    .map_err(|_| Error::InvalidPublicInputBytes),
-
-                JUBJUB_SCALAR => JubJubScalar::from_bytes(&array_bytes)
-                    .map(|s| Self::JubJubScalar(s, 0))
-                    .map_err(|_| Error::InvalidPublicInputBytes),
-
-                JUBJUB_AFFINE => JubJubAffine::from_bytes(&array_bytes)
-                    .map(|s| Self::AffinePoint(s, 0, 0))
-                    .map_err(|_| Error::InvalidPublicInputBytes),
-
-                _ => unreachable!(),
-            }
-        }
+impl From<JubJubScalar> for PublicInputValue {
+    fn from(scalar: JubJubScalar) -> Self {
+        Self(vec![scalar.into()])
     }
+}
 
-    /// Returns the positions that of a PublicInput struct
-    fn pos(&self) -> [usize; 2] {
-        match self {
-            PublicInput::BlsScalar(_, pos) => [*pos, 0],
-            PublicInput::JubJubScalar(_, pos) => [*pos, 0],
-            PublicInput::AffinePoint(_, pos_x, pos_y) => [*pos_x, *pos_y],
-        }
+impl From<JubJubAffine> for PublicInputValue {
+    fn from(point: JubJubAffine) -> Self {
+        Self(vec![point.get_x(), point.get_y()])
     }
 }
 
+type PublicInputPositions = Vec<usize>;
+
 /// Circuit representation for a gadget with all of the tools that it
 /// should implement.
 pub trait Circuit<'a>
 where
     Self: Sized,
 {
+    /// Initialization string used to fill the transcript for both parties.
+    const TRANSCRIPT_INIT: &'static [u8];
+    /// Trimming size for the keys of the circuit.
+    const TRIM_SIZE: usize;
```

Comment:
`TRIM_SIZE` can't be constant for a circuit. One example is the transfer circuit that has variable circuit size depending on the number of inputs

Reply:
I thought about it when I was implementing it. 
What if we pass it with a const generic? SO that the trim size also constrains the type @ZER0 @vlopes11 ?

Reply:
Let's talk about it tomorrow morning. It's crucial to fully understand the edge case of transfer circuit, it's the only anomaly there.

Reply:
#351 is only about the `set_trim_size` that should never be called, as in the proposal inside the issue

> Proposal: remove `Circuit::set_trim_size`

However, the `get_trim_size` (or any other method name with the same effect) will return the current size for that circuit. There is the "con" to tie the circuit instance to a fixed size trim without any clear "pro"